[JSC] Fix ASSERT(canSafelyBeUsed()) failure in Waiter's WeakPtr<TicketData>::get() with ThreadSafeWeakPtr

hyjorc1 opened the pull request:
#### 36e90619cf8a0f01a4cf8f0d414140c78410279b
<pre>
[JSC] Fix ASSERT(canSafelyBeUsed()) failure in Waiter&apos;s WeakPtr&lt;TicketData&gt;::get() with ThreadSafeWeakPtr
<a href="https://bugs.webkit.org/show_bug.cgi?id=276527">https://bugs.webkit.org/show_bug.cgi?id=276527</a>
<a href="https://rdar.apple.com/131590927">rdar://131590927</a>

Reviewed by Keith Miller.

The WeakPtr&lt;TicketData&gt; in the Waiter of WaiterListManager can be
accessed from various threads. Most of the accesses to the TicketData
are guarded by the corresponding WaiterList&apos;s lock. However,
DeferredWorkTimer::cancelPendingWorkSafe can force the cancellation
of the pending TicketData during the destruction of JSGlobalObject.
Therefore, let&apos;s use ThreadSafeWeakPtr to secure the usage of the Waiter&apos;s TicketData.

This patch enhances the robustness of WaiterListManager by:
1. Replacing WeakPtr with ThreadSafeWeakPtr for the Waiter&apos;s TicketData.
2. Refactoring scheduleWorkAndClearTicket to scheduleWorkAndClear.
3. Refactoring cancelAsyncWaiter to cancelAndClear.

* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::addWeakTicket):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::Waiter::scheduleWorkAndClear):
(JSC::Waiter::cancelAndClear):
(JSC::WaiterListManager::unregister):
(JSC::Waiter::dump const):
(JSC::Waiter::~Waiter): Deleted.
(JSC::Waiter::scheduleWorkAndClearTicket): Deleted.
(JSC::WaiterListManager::cancelAsyncWaiter): Deleted.
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: <a href="https://commits.webkit.org/280909@main">https://commits.webkit.org/280909@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5fd6aaf3dd61c82abee84b483413243deb594d43

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58002 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10479 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61627 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8447 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44966 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8636 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/6018 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60032 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/35008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/50137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/27839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/31772 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/7432 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7451 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/51094 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/7700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63315 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/57244 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1916 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/7770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54231 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1923 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/50148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/54371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/88/builds/1641 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/79005 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8650 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/33159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/13109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34245 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35329 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33990 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/WaiterListManager.cpp:
```diff
@@ -275,13 +255,25 @@ size_t WaiterListManager::totalWaiterCount()
+    ASSERT(m_isAsync);
+    if (auto ticket = this->ticket(listLocker)) {
+        m_vm->deferredWorkTimer->cancelPendingWork(ticket.get());
+        m_vm->deferredWorkTimer->scheduleWorkSoon(ticket.get(), [](DeferredWorkTimer::Ticket) mutable { });
```
Why does the lambda need to be mutable? Can you remind me why we need to schedule work when we clear? Why does cancelPendingWork not clear the ticket now that it can be `ThreadSafeWeakPtr`ed?

hyjorc1 replied:
> Why does the lambda need to be mutable?

No, we don't need it. I directly copied it from the previous use of scheduling an empty task. Will change.

> Can you remind me why we need to schedule work when we clear?

`cancelPendingWork` just marks the TicketData in the canceled status. Then, `scheduleWorkSoon` will fire the work with the TicketData and remove it from the `HashSet<Ref<TicketData>> m_pendingTickets` in `DeferredWorkTimer`. Here, we just explicitly schedules an empty work after the cancelation.

> Why does cancelPendingWork not clear the ticket now that it can be ThreadSafeWeakPtred?

Yeah, I think we can do that. Since TicketData is `ThreadSafeWeakPtr` now, we can definitely optimize the workflow in the current `DeferredWorkTimer`. Will add a `FIXME` with a bugzilla link.

hyjorc1 replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=276538 [JSC] Optimize the workflow in DeferredWorkTimer since TicketData is ThreadSafeWeakPtr now

kmiller68 commented on Source/JavaScriptCore/runtime/WaiterListManager.h:
```diff
@@ -90,21 +93,25 @@ class Waiter final : public WTF::BasicRawSentinelNode<Waiter>, public ThreadSafe
         m_timer = nullptr;
     }
 
+    void scheduleWorkAndClear(const AbstractLocker&, DeferredWorkTimer::Task&&);
+    void cancelAndClear(const AbstractLocker&);
```
We should use Clang thread safety analysis instead of explicitly passing the lock. I guess it's a bit tricky because the lock is in the Manager? I'm not sure if there's a good way to express that... @kkinnunen-apple @cdumez Do you know?

hyjorc1 replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=276539 [JSC] Use Clang thread safety analysis for scheduleWorkAndClear and cancelAndClear in WaiterListManager.

kmiller68 reviewed (approved):
r=me with future ideas.

hyjorc1 committed:
[JSC] Fix ASSERT(canSafelyBeUsed()) failure in Waiter's WeakPtr<TicketData>::get() with ThreadSafeWeakPtr
https://bugs.webkit.org/show_bug.cgi?id=276527
rdar://131590927

Reviewed by Keith Miller.

The WeakPtr<TicketData> in the Waiter of WaiterListManager can be
accessed from various threads. Most of the accesses to the TicketData
are guarded by the corresponding WaiterList's lock. However,
DeferredWorkTimer::cancelPendingWorkSafe can force the cancellation
of the pending TicketData during the destruction of JSGlobalObject.
Therefore, let's use ThreadSafeWeakPtr to secure the usage of the Waiter's TicketData.

This patch enhances the robustness of WaiterListManager by:
1. Replacing WeakPtr with ThreadSafeWeakPtr for the Waiter's TicketData.
2. Refactoring scheduleWorkAndClearTicket to scheduleWorkAndClear.
3. Refactoring cancelAsyncWaiter to cancelAndClear.

* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::addWeakTicket):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/WaiterListManager.cpp:
(JSC::WaiterListManager::notifyWaiterImpl):
(JSC::Waiter::scheduleWorkAndClear):
(JSC::Waiter::cancelAndClear):
(JSC::WaiterListManager::unregister):
(JSC::Waiter::dump const):
(JSC::Waiter::~Waiter): Deleted.
(JSC::Waiter::scheduleWorkAndClearTicket): Deleted.
(JSC::WaiterListManager::cancelAsyncWaiter): Deleted.
* Source/JavaScriptCore/runtime/WaiterListManager.h:

Canonical link: https://commits.webkit.org/280909@main
